pkg/controller/persistentvolume/state.go

```diff
@@ -0,0 +1,307 @@
+	case api.VolumeFailed:
+		ctx.state = &persistentVolumeFailedState{}
+	}
+	ctx.client = client
+	ctx.volumeIndex = volumeIndex
+	ctx.volume = pv
+
+	if err := ctx.fetchClaim(); err != nil {
+		return nil, err
+	}
+	return ctx, nil
+}
+
+// fetchClaim will find the claim bound to the PV, if any, and set it on the persistentVolumeContext object.
+// any error that is not 404 will be returned.  Missing claims (404s) are normal and not fatal to the PV's state.
+// Claims that 404 are important to track, because a deleted claim is meaningful to a volume bound to it.
+func (ctx *persistentVolumeStateContext) fetchClaim() error {
+	if ctx.volume != nil && ctx.volume.Spec.ClaimRef != nil {
+		claim, err := ctx.client.GetPersistentVolumeClaim(ctx.volume.Spec.ClaimRef.Namespace, ctx.volume.Spec.ClaimRef.Name)
+		if err != nil && !errors.IsNotFound(err) {
+			return err
+		}
+		if err != nil && errors.IsNotFound(err) {
+			ctx.claimNotFound = true
+			return nil
+		}
+		ctx.claim = claim
+		ctx.claimNotFound = false
+	}
+	return nil
+}
+
+func (c *persistentVolumeStateContext) saveStatus() error {
+	return c.state.saveStatus(c)
+}
+
+func (c *persistentVolumeStateContext) bind() error {
+	return c.state.bind(c)
+}
+
+func (c *persistentVolumeStateContext) recycle() error {
+	return c.state.recycle(c)
+}
+
+func (c *persistentVolumeStateContext) delete() error {
+	return c.state.delete(c)
+}
+
+func (c *persistentVolumeStateContext) requiresProvisioning() bool {
+	return c.state.requiresProvisioning(c)
+}
+
+func (c *persistentVolumeStateContext) provision() error {
+	return c.state.provision(c)
+}
+
+func (c *persistentVolumeStateContext) isProvisioningComplete() bool {
+	return c.state.isProvisioningComplete(c)
+}
+
+// STATE OBJECTS
+
+type persistentVolumePendingState struct{}
+type persistentVolumeAvailableState struct{}
+type persistentVolumeBoundState struct{}
+type persistentVolumeReleasedState struct{}
+type persistentVolumeFailedState struct{}
+
+// PENDING
+
+func (state *persistentVolumePendingState) saveStatus(ctx *persistentVolumeStateContext) error {
+	fmt.Printf("next phase is available, maybe keep pending")
+	return nil
+}
+
+func (state *persistentVolumePendingState) bind(ctx *persistentVolumeStateContext) error {
+	return fmt.Errorf("volumePendingState is the wrong state for `bind`")
+}
+
+func (state *persistentVolumePendingState) recycle(ctx *persistentVolumeStateContext) error {
+	return fmt.Errorf("volumePendingState is the wrong state for `recycle`")
+}
+
+func (state *persistentVolumePendingState) delete(ctx *persistentVolumeStateContext) error {
+	return fmt.Errorf("volumePendingState is the wrong state for `delete`")
+}
+
+func (state *persistentVolumePendingState) requiresProvisioning(ctx *persistentVolumeStateContext) bool {
+	if !keyExists(pvProvisioningRequiredAnnotationKey, ctx.volume.Annotations) {
+		return false
+	}
+	if isProvisioningComplete(ctx.volume) {
+		return false
+	}
+	return true
```

Comment:
I am not sure yet but something like this should probably just return false or false depending on what the state means. Another operation like setPvProvisioningRequiredAnnotationKey() would change the state of the object to where the implementation of requiresProvisioning returns always true or always false depending on the meaning. Sorry I cannot be more precise... I am missing some context of what the different states mean and what the differences between them